// @angular-material-components/file-input/esm2020/lib/file-input.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatLegacyButtonModule as MatButtonModule } from '@angular/material/legacy-button';
import { MatLegacyFormFieldModule as MatFormFieldModule } from '@angular/material/legacy-form-field';
import { MatIconModule } from '@angular/material/icon';
import { MatLegacyInputModule as MatInputModule } from '@angular/material/legacy-input';
import { NgxMatFileInputComponent, NgxMatFileInputIcon } from './file-input.component';
import * as i0 from "@angular/core";
export class NgxMatFileInputModule {
}
/** @nocollapse */ NgxMatFileInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: NgxMatFileInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ NgxMatFileInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.2", ngImport: i0, type: NgxMatFileInputModule, declarations: [NgxMatFileInputComponent,
        NgxMatFileInputIcon], imports: [CommonModule,
        MatButtonModule,
        MatIconModule,
        MatFormFieldModule,
        MatInputModule], exports: [NgxMatFileInputComponent,
        NgxMatFileInputIcon] });
/** @nocollapse */ NgxMatFileInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: NgxMatFileInputModule, imports: [CommonModule,
        MatButtonModule,
        MatIconModule,
        MatFormFieldModule,
        MatInputModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: NgxMatFileInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxMatFileInputComponent,
                        NgxMatFileInputIcon
                    ],
                    imports: [
                        CommonModule,
                        MatButtonModule,
                        MatIconModule,
                        MatFormFieldModule,
                        MatInputModule
                    ],
                    exports: [
                        NgxMatFileInputComponent,
                        NgxMatFileInputIcon
                    ]
                }]
        }] });
